pkg/ingester/stream.go

```diff
@@ -193,6 +198,32 @@ func (s *stream) Push(_ context.Context, entries []logproto.Entry) error {
 	return nil
 }
 
+// Returns true, if chunk should be cut before adding new entry. This is done to make ingesters
+// cut the chunk for this stream at the same moment, so that new chunk will contain exactly the same entries.
+func (s *stream) cutChunkForSynchronization(entryTimestamp, prevEntryTimestamp time.Time, chunk chunkenc.Chunk, synchronizePeriod time.Duration, minUtilization float64) bool {
+	if synchronizePeriod <= 0 || prevEntryTimestamp.IsZero() {
+		return false
+	}
+
+	// we use fingerprint as a jitter here, basically offsetting stream synchronization points to different
+	// this breaks if streams are mapped to different fingerprints on different ingesters, which is too bad.
+	cts := (uint64(entryTimestamp.UnixNano()) + uint64(s.fp)) % uint64(synchronizePeriod.Nanoseconds())
+	pts := (uint64(prevEntryTimestamp.UnixNano()) + uint64(s.fp)) % uint64(synchronizePeriod.Nanoseconds())
+
+	// if current entry timestamp has rolled over synchronization period
+	if cts < pts {
```

Comment:
This seems incorrect to me?  Assuming entryTimestamp is always greater than the prevEntryTimestamp (as is required that all logs have increasing timestamps) this comparison can never be true?

prev = 10, current = 20 

20 + fp + sync will never be less than 10 + fp + sync?

Reply:
err, sorry, it's not fp + sync it's fp % sync however the result would be the same for both equations:

20 + (fp % sync) will never be less than 10 + (fp % sync)

Reply:
you're reading it wrong. it's `(timestamp + fp) % sync`

it will be true when it rolls over the sync period. See the new test and output it produces.

Reply:
those extra parenthesis for casting make it unreadable :(

Reply:
> it will be true when it rolls over the sync period

that is old timestamp is in old sync period, new one is in new period. Modulo arithmetics will make new timestamp smaller than old one.

Reply:
ah! ok that makes more sense!

Reply:
```
    instance_test.go:140: 17:19:37.737 -- 17:20:59.098 1m21.360836928s 0.2755859375
    instance_test.go:140: 17:20:59.691 -- 17:22:58.799 1m59.107617256s 0.421875
    instance_test.go:140: 17:22:59.338 -- 17:23:58.699 59.360134469s 0.214453125
    instance_test.go:140: 17:23:59.300 -- 17:24:59.066 59.766629179s 0.2126953125
    instance_test.go:140: 17:24:59.931 -- 17:25:58.866 58.934245888s 0.20390625
    instance_test.go:140: 17:25:59.838 -- 17:26:59.096 59.258072082s 0.20390625
    instance_test.go:140: 17:26:59.510 -- 17:27:57.717 58.207103008s 0.20390625
```

This is with 10min period and 0.2 min util.